jlm/llvm/opt/ScalarEvolution.cpp

```diff
@@ -0,0 +1,209 @@
+#include <iostream>
+#include <jlm/llvm/ir/operators/IntegerOperations.hpp>
+#include <jlm/llvm/opt/ScalarEvolution.hpp>
+#include <jlm/rvsdg/gamma.hpp>
+#include <jlm/rvsdg/lambda.hpp>
+#include <jlm/rvsdg/MatchType.hpp>
+#include <jlm/rvsdg/RvsdgModule.hpp>
+#include <jlm/rvsdg/theta.hpp>
+#include <jlm/rvsdg/traverser.hpp>
+#include <jlm/util/Statistics.hpp>
+
+#include <variant>
+
+namespace jlm::llvm
+{
+
+class ScalarEvolution::Statistics final : public util::Statistics
+{
+
+public:
+  ~Statistics() override = default;
+
+  explicit Statistics(const util::FilePath & sourceFile)
+      : util::Statistics(Id::ScalarEvolution, sourceFile)
+  {}
+
+  void
+  Start() noexcept
+  {
+    AddTimer(Label::Timer).start();
+  }
+
+  void
+  Stop() noexcept
+  {
+    GetTimer(Label::Timer).stop();
+  }
+
+  static std::unique_ptr<Statistics>
+  Create(const util::FilePath & sourceFile)
+  {
+    return std::make_unique<Statistics>(sourceFile);
+  }
+};
+
+void
+ScalarEvolution::Run(
+    rvsdg::RvsdgModule & rvsdgModule,
+    util::StatisticsCollector & statisticsCollector)
+{
+  auto statistics = Statistics::Create(rvsdgModule.SourceFilePath().value());
+  statistics->Start();
```

Comment:
InductionVariableMap_ needs to be cleared at the beginning of this function. It should be possible to invoke the Run() method multiple times from the same instance.